File: apps/app/src/lib/comparative-stats.ts
import { z } from 'zod'
import type { IndicatorData } from '@/lib/analyze-shared'

/**
 * Cross-asset statistics for the multi-token analysis: computed CLIENT-SIDE
 * from the daily price histories (LLMs can't be trusted to do this math),
 * displayed in the dialog's comparative sidebar, and passed to
 * /api/analyze/compare as precomputed ground truth for the prompt.
 *
 * All series are daily closes (1 point = 1 day), most-recent last.
 */

export const ComparativeTokenStatsSchema = z.object({
  id: z.string(),
  symbol: z.string(),
  name: z.string(),
  marketCap: z.number().nullable(),
  /** % return over the last 7 daily closes. */
  return7dPct: z.number().nullable(),
  /** % return over the provided history (~30d). */
  return30dPct: z.number().nullable(),
  /** 7d return minus the benchmark's 7d return (percentage points). */
  excessReturn7dPct: z.number().nullable(),
  /** 30d return minus the benchmark's 30d return (percentage points). */
  excessReturn30dPct: z.number().nullable(),
  /** Annualized realized volatility of daily returns, in %. */
  volatility30dAnnualizedPct: z.number().nullable(),
  /** OLS beta of daily returns vs the benchmark (benchmark = 1). */
  betaVsBenchmark: z.number().nullable(),
  rsi: z.number().nullable(),
  /** Position of the RSI-Bollinger value within its bands (0=lower, 1=upper). */
  bbPercentB: z.number().nullable(),
  /** Bollinger Band Width Percentile (0-100) on daily closes; <20 = squeeze, >80 = expansion climax. */
  bbwpPct: z.number().nullable(),
  /** Wave Trend posture, e.g. "bullish_cross (strong)". */
  waveTrend: z.string().nullable(),
  /** Money Flow posture, e.g. "inflow (moderate)". */
  moneyFlow: z.string().nullable(),
  openInterestChangePct: z.number().nullable(),
  /** 0..1 taker buy ratio. */
  takerBuyRatio: z.number().nullable(),
})

export const ComparativeStatsSchema = z.object({
  benchmarkId: z.string(),
  benchmarkSymbol: z.string(),
  tokens: z.array(ComparativeTokenStatsSchema),
  /**
   * Pearson correlation of aligned daily returns; row/col order matches
   * `tokens`. Diagonal is 1; null where the overlap is too short.
   */
  correlationMatrix: z.array(z.array(z.number().nullable())),
})

export type ComparativeTokenStats = z.infer<typeof ComparativeTokenStatsSchema>
export type ComparativeStats = z.infer<typeof ComparativeStatsSchema>

const MIN_OVERLAP_RETURNS = 10

function dailyReturns(prices: number[]): number[] {
  const returns: number[] = []
  for (let i = 1; i < prices.length; i++) {
    const prev = prices[i - 1]!
    const curr = prices[i]!
    if (prev > 0 && Number.isFinite(prev) && Number.isFinite(curr)) {
      returns.push(curr / prev - 1)
    }
  }
  return returns
}

function mean(values: number[]): number {
  return values.reduce((a, b) => a + b, 0) / values.length
}

/** Align two return series on their most recent `n = min(len)` points. */
function alignTails(a: number[], b: number[]): [number[], number[]] {
  const n = Math.min(a.length, b.length)
  return [a.slice(-n), b.slice(-n)]
}

export function pearsonCorrelation(a: number[], b: number[]): number | null {
  const [x, y] = alignTails(a, b)
  if (x.length < MIN_OVERLAP_RETURNS) return null
  const mx = mean(x)
  const my = mean(y)
  let cov = 0
  let vx = 0
  let vy = 0
  for (let i = 0; i < x.length; i++) {
    const dx = x[i]! - mx
    const dy = y[i]! - my
    cov += dx * dy
    vx += dx * dx
    vy += dy * dy
  }
  if (vx === 0 || vy === 0) return null
  return cov / Math.sqrt(vx * vy)
}

export function betaVs(token: number[], benchmark: number[]): number | null {
  const [x, y] = alignTails(token, benchmark) // x = token, y = benchmark
  if (x.length < MIN_OVERLAP_RETURNS) return null
  const mx = mean(x)
  const my = mean(y)
  let cov = 0
  let varB = 0
  for (let i = 0; i < x.length; i++) {
    cov += (x[i]! - mx) * (y[i]! - my)
    varB += (y[i]! - my) ** 2
  }
  if (varB === 0) return null
  return cov / varB
}

export function annualizedVolatilityPct(returns: number[]): number | null {
  if (returns.length < MIN_OVERLAP_RETURNS) return null
  const m = mean(returns)
  const variance =
    returns.reduce((sum, r) => sum + (r - m) ** 2, 0) / (returns.length - 1)
  return Math.sqrt(variance) * Math.sqrt(365) * 100
}

function periodReturnPct(prices: number[], periods: number): number | null {
  if (prices.length < periods + 1) {
    // Fall back to full available history for the "30d" horizon.
    if (prices.length >= 2 && periods >= 30) {
      const first = prices[0]!
      const last = prices[prices.length - 1]!
      return first > 0 ? (last / first - 1) * 100 : null
    }
    return null
  }
  const start = prices[prices.length - 1 - periods]!
  const end = prices[prices.length - 1]!
  return start > 0 ? (end / start - 1) * 100 : null
}

/** Benchmark rule: BTC if present, else ETH, else largest market cap. */
export function pickBenchmarkIndex(
  tokens: Array<{ symbol: string; marketCap: number | null }>,
): number {
  const bySymbol = (sym: string) =>
    tokens.findIndex((t) => t.symbol.toLowerCase() === sym)
  const btc = bySymbol('btc')
  if (btc >= 0) return btc
  const eth = bySymbol('eth')
  if (eth >= 0) return eth
  let best = 0
  for (let i = 1; i < tokens.length; i++) {
    if ((tokens[i]!.marketCap ?? 0) > (tokens[best]!.marketCap ?? 0)) best = i
  }
  return best
}

export interface ComparativeStatsInput {
  id: string
  data: IndicatorData
  /**
   * Timestamped chart series (unix-second `time`). Strongly preferred over
   * `data.priceContext.priceHistory`: the chart hook serves sub-daily bars
   * whose cache windows differ per token, so index-aligned histories can be
   * offset in time — timestamps let us bucket to daily closes and align
   * pairs BY DAY, which is what makes correlation/beta meaningful.
   */
  series?: Array<{ time: unknown; value: number }>
  /** Precomputed BBWP (0-100) — computed client-side from the daily closes. */
  bbwpPct?: number | null
}

interface DailyCloses {
  /** Unix day numbers (floor(unixSeconds / 86400)), ascending. */
  days: number[]
  closes: number[]
}

export function toDailyCloses(
  series: Array<{ time: unknown; value: number }>,
): DailyCloses {
  const byDay = new Map<number, number>()
  for (const point of series) {
    const t = Number(point.time)
    if (!Number.isFinite(t) || !Number.isFinite(point.value)) continue
    // Last value seen for a day wins (series is chronological → daily close).
    byDay.set(Math.floor(t / 86400), point.value)
  }
  const days = [...byDay.keys()].sort((a, b) => a - b)
  return { days, closes: days.map((d) => byDay.get(d)!) }
}

/** Fallback when no timestamps exist: treat indexes as consecutive days. */
function syntheticDailyCloses(prices: number[]): DailyCloses {
  return { days: prices.map((_, i) => i), closes: [...prices] }
}

/**
 * Returns computed over the two tokens' COMMON days only, so each return pair
 * covers the identical time span for both assets.
 */
function alignedReturnsByDay(a: DailyCloses, b: DailyCloses): [number[], number[]] {
  const bByDay = new Map<number, number>()
  b.days.forEach((day, i) => bByDay.set(day, b.closes[i]!))
  const closesA: number[] = []
  const closesB: number[] = []
  a.days.forEach((day, i) => {
    const closeB = bByDay.get(day)
    if (closeB !== undefined) {
      closesA.push(a.closes[i]!)
      closesB.push(closeB)
    }
  })
  return [dailyReturns(closesA), dailyReturns(closesB)]
}

const CORRELATION_WINDOW_DAYS = 30

/** Build the full cross-asset stats block from each token's data. */
export function computeComparativeStats(
  tokens: ComparativeStatsInput[],
): ComparativeStats {
  const bases = tokens.map(({ id, data, series, bbwpPct }) => {
    const daily =
      series && series.length > 0
        ? toDailyCloses(series)
        : syntheticDailyCloses(data.priceContext?.priceHistory ?? [])
    const bb = data.bollingerBands
    const bandWidth = bb ? bb.upperBand - bb.lowerBand : 0
    const wt = data.marketVision?.waveTrend
    const mf = data.marketVision?.moneyFlow
    return {
      id,
      symbol: data.symbol,
      name: data.name,
      marketCap: data.quote.USD.market_cap ?? null,
      daily,
      rsi: data.marketVision?.rsi?.value ?? null,
      bbPercentB:
        bb && bandWidth > 0 ? (bb.currentValue - bb.lowerBand) / bandWidth : null,
      bbwpPct: bbwpPct ?? null,
      waveTrend: wt ? `${wt.signal}${wt.momentum ? ` (${wt.momentum})` : ''}` : null,
      moneyFlow: mf ? `${mf.direction} (${mf.strength})` : null,
      openInterestChangePct: data.liquidationData?.openInterestChange ?? null,
      takerBuyRatio: data.orderFlow?.takerBuyRatio ?? null,
    }
  })

  const benchIdx = pickBenchmarkIndex(bases)
  const bench = bases[benchIdx]!
  const benchReturn7d = periodReturnPct(bench.daily.closes, 7)
  const benchReturn30d = periodReturnPct(bench.daily.closes, 30)

  const tokenStats: ComparativeTokenStats[] = bases.map((base, i) => {
    const return7dPct = periodReturnPct(base.daily.closes, 7)
    const return30dPct = periodReturnPct(base.daily.closes, 30)
    const [tokenReturns, benchReturns] = alignedReturnsByDay(
      base.daily,
      bench.daily,
    )
    return {
      id: base.id,
      symbol: base.symbol,
      name: base.name,
      marketCap: base.marketCap,
      return7dPct,
      return30dPct,
      excessReturn7dPct:
        return7dPct !== null && benchReturn7d !== null
          ? return7dPct - benchReturn7d
          : null,
      excessReturn30dPct:
        return30dPct !== null && benchReturn30d !== null
          ? return30dPct - benchReturn30d
          : null,
      volatility30dAnnualizedPct: annualizedVolatilityPct(
        dailyReturns(base.daily.closes).slice(-CORRELATION_WINDOW_DAYS),
      ),
      betaVsBenchmark:
        i === benchIdx
          ? 1
          : betaVs(
              tokenReturns.slice(-CORRELATION_WINDOW_DAYS),
              benchReturns.slice(-CORRELATION_WINDOW_DAYS),
            ),
      rsi: base.rsi,
      bbPercentB: base.bbPercentB,
      bbwpPct: base.bbwpPct,
      waveTrend: base.waveTrend,
      moneyFlow: base.moneyFlow,
      openInterestChangePct: base.openInterestChangePct,
      takerBuyRatio: base.takerBuyRatio,
    }
  })

  const correlationMatrix = bases.map((a, i) =>
    bases.map((b, j) => {
      if (i === j) return 1
      const [ra, rb] = alignedReturnsByDay(a.daily, b.daily)
      return pearsonCorrelation(
        ra.slice(-CORRELATION_WINDOW_DAYS),
        rb.slice(-CORRELATION_WINDOW_DAYS),
      )
    }),
  )

  return {
    benchmarkId: bench.id,
    benchmarkSymbol: bench.symbol,
    tokens: tokenStats,
    correlationMatrix,
  }
}

const fmt = (v: number | null, digits = 2, suffix = ''): string =>
  v === null ? 'N/A' : `${v.toFixed(digits)}${suffix}`

/** Markdown block for the compare prompt — precomputed ground truth. */
export function formatComparativeStats(stats: ComparativeStats): string {
  const lines: string[] = []
  lines.push(
    `Benchmark: ${stats.benchmarkSymbol.toUpperCase()} (rule: BTC, else ETH, else largest market cap in the selection)`,
  )
  lines.push('')
  lines.push(
    '| Asset | 7d Return | 30d Return | Excess 7d vs Bench | Excess 30d vs Bench | Ann. Volatility | Beta vs Bench | RSI | OI Δ | Taker Buy |',
  )
  lines.push('|---|---|---|---|---|---|---|---|---|---|')
  for (const t of stats.tokens) {
    lines.push(
      `| ${t.symbol.toUpperCase()} | ${fmt(t.return7dPct, 2, '%')} | ${fmt(t.return30dPct, 2, '%')} | ${fmt(t.excessReturn7dPct, 2, 'pp')} | ${fmt(t.excessReturn30dPct, 2, 'pp')} | ${fmt(t.volatility30dAnnualizedPct, 0, '%')} | ${fmt(t.betaVsBenchmark)} | ${fmt(t.rsi, 1)} | ${fmt(t.openInterestChangePct, 2, '%')} | ${t.takerBuyRatio === null ? 'N/A' : `${(t.takerBuyRatio * 100).toFixed(1)}%`} |`,
    )
  }
  lines.push('')
  lines.push('Indicator posture per asset (Wave Trend / Money Flow from Market Vision; %B = RSI-Bollinger position 0..1; BBWP = Bollinger Band Width Percentile on daily closes, <20 squeeze / >80 expansion climax):')
  lines.push('| Asset | Wave Trend | Money Flow | RSI-BB %B | BBWP |')
  lines.push('|---|---|---|---|---|')
  for (const t of stats.tokens) {
    lines.push(
      `| ${t.symbol.toUpperCase()} | ${t.waveTrend ?? 'N/A'} | ${t.moneyFlow ?? 'N/A'} | ${fmt(t.bbPercentB, 2)} | ${fmt(t.bbwpPct, 0)} |`,
    )
  }
  lines.push('')
  const symbols = stats.tokens.map((t) => t.symbol.toUpperCase())
  lines.push('Pairwise correlation of daily returns (30d):')
  lines.push(`| | ${symbols.join(' | ')} |`)
  lines.push(`|---|${symbols.map(() => '---').join('|')}|`)
  stats.correlationMatrix.forEach((row, i) => {
    lines.push(
      `| ${symbols[i]} | ${row.map((v) => (v === null ? 'N/A' : v.toFixed(2))).join(' | ')} |`,
    )
  })
  return lines.join('\n')
}

File: apps/ios/Packages/AggrCore/Sources/AggrCore/Analysis/ComparativeStats.swift
import Foundation

/// Port of `lib/comparative-stats.ts`: cross-asset stats computed CLIENT-SIDE from daily closes.
public enum ComparativeStats {
  static let minOverlapReturns = 10
  static let correlationWindowDays = 30

  public struct Input: Sendable {
    public var id: String
    public var symbol: String
    public var name: String
    public var marketCap: Double?
    /// Timestamped series (epoch seconds) — bucketed to daily closes.
    public var series: [TimePoint]
    /// Fallback index-aligned closes when no timestamps exist.
    public var priceHistory: [Double]
    public var rsi: Double?
    public var bbPercentB: Double?
    public var bbwpPct: Double?
    public var waveTrend: String?
    public var moneyFlow: String?
    public var openInterestChangePct: Double?
    public var takerBuyRatio: Double?

    public init(id: String, symbol: String, name: String, marketCap: Double?, series: [TimePoint] = [], priceHistory: [Double] = [],
                rsi: Double? = nil, bbPercentB: Double? = nil, bbwpPct: Double? = nil, waveTrend: String? = nil, moneyFlow: String? = nil,
                openInterestChangePct: Double? = nil, takerBuyRatio: Double? = nil) {
      self.id = id; self.symbol = symbol; self.name = name; self.marketCap = marketCap; self.series = series; self.priceHistory = priceHistory
      self.rsi = rsi; self.bbPercentB = bbPercentB; self.bbwpPct = bbwpPct; self.waveTrend = waveTrend; self.moneyFlow = moneyFlow
      self.openInterestChangePct = openInterestChangePct; self.takerBuyRatio = takerBuyRatio
    }
  }

  public struct TokenStats: Sendable, Hashable, Codable, Identifiable {
    public var id: String
    public var symbol: String
    public var name: String
    public var marketCap: Double?
    public var return7dPct: Double?
    public var return30dPct: Double?
    public var excessReturn7dPct: Double?
    public var excessReturn30dPct: Double?
    public var volatility30dAnnualizedPct: Double?
    public var betaVsBenchmark: Double?
    public var rsi: Double?
    public var bbPercentB: Double?
    public var bbwpPct: Double?
    public var waveTrend: String?
    public var moneyFlow: String?
    public var openInterestChangePct: Double?
    public var takerBuyRatio: Double?
  }

  public struct Result: Sendable, Hashable, Codable {
    public var benchmarkId: String
    public var benchmarkSymbol: String
    public var tokens: [TokenStats]
    /// Row/col order matches `tokens`; nil where overlap is too short.
    public var correlationMatrix: [[Double?]]
  }

  public struct DailyCloses: Sendable { public var days: [Int]; public var closes: [Double] }

  static func dailyReturns(_ prices: [Double]) -> [Double] {
    var out: [Double] = []
    for i in 1..<max(1, prices.count) {
      let prev = prices[i - 1], curr = prices[i]
      if prev > 0, prev.isFinite, curr.isFinite { out.append(curr / prev - 1) }
    }
    return out
  }

  static func mean(_ v: [Double]) -> Double { v.reduce(0, +) / Double(v.count) }

  static func alignTails(_ a: [Double], _ b: [Double]) -> ([Double], [Double]) {
    let n = min(a.count, b.count)
    return (Array(a.suffix(n)), Array(b.suffix(n)))
  }

  public static func pearsonCorrelation(_ a: [Double], _ b: [Double]) -> Double? {
    let (x, y) = alignTails(a, b)
    guard x.count >= minOverlapReturns else { return nil }
    let mx = mean(x), my = mean(y)
    var cov = 0.0, vx = 0.0, vy = 0.0
    for i in 0..<x.count { let dx = x[i] - mx, dy = y[i] - my; cov += dx * dy; vx += dx * dx; vy += dy * dy }
    guard vx != 0, vy != 0 else { return nil }
    return cov / (vx * vy).squareRoot()
  }

  public static func betaVs(_ token: [Double], _ benchmark: [Double]) -> Double? {
    let (x, y) = alignTails(token, benchmark)
    guard x.count >= minOverlapReturns else { return nil }
    let mx = mean(x), my = mean(y)
    var cov = 0.0, varB = 0.0
    for i in 0..<x.count { cov += (x[i] - mx) * (y[i] - my); varB += (y[i] - my) * (y[i] - my) }
    guard varB != 0 else { return nil }
    return cov / varB
  }

  public static func annualizedVolatilityPct(_ returns: [Double]) -> Double? {
    guard returns.count >= minOverlapReturns else { return nil }
    let m = mean(returns)
    let variance = returns.reduce(0) { $0 + ($1 - m) * ($1 - m) } / Double(returns.count - 1)
    return variance.squareRoot() * (365.0).squareRoot() * 100
  }

  static func periodReturnPct(_ prices: [Double], periods: Int) -> Double? {
    if prices.count < periods + 1 {
      if prices.count >= 2, periods >= 30 { let f = prices[0], l = prices[prices.count - 1]; return f > 0 ? (l / f - 1) * 100 : nil }
      return nil
    }
    let start = prices[prices.count - 1 - periods], end = prices[prices.count - 1]
    return start > 0 ? (end / start - 1) * 100 : nil
  }

  /// Benchmark rule: BTC if present, else ETH, else largest market cap.
  public static func pickBenchmarkIndex(_ tokens: [(symbol: String, marketCap: Double?)]) -> Int {
    if let i = tokens.firstIndex(where: { $0.symbol.lowercased() == "btc" }) { return i }
    if let i = tokens.firstIndex(where: { $0.symbol.lowercased() == "eth" }) { return i }
    var best = 0
    for i in 1..<max(1, tokens.count) where (tokens[i].marketCap ?? 0) > (tokens[best].marketCap ?? 0) { best = i }
    return best
  }

  public static func toDailyCloses(_ series: [TimePoint]) -> DailyCloses {
    var byDay: [Int: Double] = [:]
    for p in series where p.value.isFinite { byDay[Int((Double(p.epochSeconds) / 86_400).rounded(.down))] = p.value }
    let days = byDay.keys.sorted()
    return DailyCloses(days: days, closes: days.map { byDay[$0]! })
  }

  static func syntheticDailyCloses(_ prices: [Double]) -> DailyCloses { DailyCloses(days: Array(0..<prices.count), closes: prices) }

  static func alignedReturnsByDay(_ a: DailyCloses, _ b: DailyCloses) -> ([Double], [Double]) {
    let bByDay = Dictionary(zip(b.days, b.closes), uniquingKeysWith: { _, n in n })
    var ca: [Double] = [], cb: [Double] = []
    for (i, day) in a.days.enumerated() { if let closeB = bByDay[day] { ca.append(a.closes[i]); cb.append(closeB) } }
    return (dailyReturns(ca), dailyReturns(cb))
  }

  public static func compute(_ tokens: [Input]) -> Result? {
    guard !tokens.isEmpty else { return nil }
    struct Base { var input: Input; var daily: DailyCloses }
    let bases = tokens.map { Base(input: $0, daily: $0.series.isEmpty ? syntheticDailyCloses($0.priceHistory) : toDailyCloses($0.series)) }
    let benchIdx = pickBenchmarkIndex(bases.map { ($0.input.symbol, $0.input.marketCap) })
    let bench = bases[benchIdx]
    let bench7 = periodReturnPct(bench.daily.closes, periods: 7)
    let bench30 = periodReturnPct(bench.daily.closes, periods: 30)
    let stats: [TokenStats] = bases.enumerated().map { i, base in
      let r7 = periodReturnPct(base.daily.closes, periods: 7)
      let r30 = periodReturnPct(base.daily.closes, periods: 30)
      let (tr, br) = alignedReturnsByDay(base.daily, bench.daily)
      return TokenStats(
        id: base.input.id, symbol: base.input.symbol, name: base.input.name, marketCap: base.input.marketCap,
        return7dPct: r7, return30dPct: r30,
        excessReturn7dPct: (r7 != nil && bench7 != nil) ? r7! - bench7! : nil,
        excessReturn30dPct: (r30 != nil && bench30 != nil) ? r30! - bench30! : nil,
        volatility30dAnnualizedPct: annualizedVolatilityPct(Array(dailyReturns(base.daily.closes).suffix(correlationWindowDays))),
        betaVsBenchmark: i == benchIdx ? 1 : betaVs(Array(tr.suffix(correlationWindowDays)), Array(br.suffix(correlationWindowDays))),
        rsi: base.input.rsi, bbPercentB: base.input.bbPercentB, bbwpPct: base.input.bbwpPct, waveTrend: base.input.waveTrend,
        moneyFlow: base.input.moneyFlow, openInterestChangePct: base.input.openInterestChangePct, takerBuyRatio: base.input.takerBuyRatio)
    }
    let matrix: [[Double?]] = bases.enumerated().map { i, a in
      bases.enumerated().map { j, b in
        if i == j { return 1 }
        let (ra, rb) = alignedReturnsByDay(a.daily, b.daily)
        return pearsonCorrelation(Array(ra.suffix(correlationWindowDays)), Array(rb.suffix(correlationWindowDays)))
      }
    }
    return Result(benchmarkId: bench.input.id, benchmarkSymbol: bench.input.symbol, tokens: stats, correlationMatrix: matrix)
  }

  static func fmt(_ v: Double?, _ digits: Int = 2, _ suffix: String = "") -> String {
    guard let v else { return "N/A" }
    return String(format: "%.\(digits)f", v) + suffix
  }

  /// Markdown block for the compare prompt — precomputed ground truth.
  public static func format(_ s: Result) -> String {
    var lines: [String] = []
    lines.append("Benchmark: \(s.benchmarkSymbol.uppercased()) (rule: BTC, else ETH, else largest market cap in the selection)")
    lines.append("")
    lines.append("| Asset | 7d Return | 30d Return | Excess 7d vs Bench | Excess 30d vs Bench | Ann. Volatility | Beta vs Bench | RSI | OI Δ | Taker Buy |")
    lines.append("|---|---|---|---|---|---|---|---|---|---|")
    for t in s.tokens {
      let taker = t.takerBuyRatio.map { String(format: "%.1f%%", $0 * 100) } ?? "N/A"
      lines.append("| \(t.symbol.uppercased()) | \(fmt(t.return7dPct, 2, "%")) | \(fmt(t.return30dPct, 2, "%")) | \(fmt(t.excessReturn7dPct, 2, "pp")) | \(fmt(t.excessReturn30dPct, 2, "pp")) | \(fmt(t.volatility30dAnnualizedPct, 0, "%")) | \(fmt(t.betaVsBenchmark)) | \(fmt(t.rsi, 1)) | \(fmt(t.openInterestChangePct, 2, "%")) | \(taker) |")
    }
    lines.append("")
    lines.append("Indicator posture per asset (Wave Trend / Money Flow from Market Vision; %B = RSI-Bollinger position 0..1; BBWP = Bollinger Band Width Percentile on daily closes, <20 squeeze / >80 expansion climax):")
    lines.append("| Asset | Wave Trend | Money Flow | RSI-BB %B | BBWP |")
    lines.append("|---|---|---|---|---|")
    for t in s.tokens { lines.append("| \(t.symbol.uppercased()) | \(t.waveTrend ?? "N/A") | \(t.moneyFlow ?? "N/A") | \(fmt(t.bbPercentB, 2)) | \(fmt(t.bbwpPct, 0)) |") }
    lines.append("")
    let symbols = s.tokens.map { $0.symbol.uppercased() }
    lines.append("Pairwise correlation of daily returns (30d):")
    lines.append("| | \(symbols.joined(separator: " | ")) |")
    lines.append("|---|\(symbols.map { _ in "---" }.joined(separator: "|"))|")
    for (i, row) in s.correlationMatrix.enumerated() {
      lines.append("| \(symbols[i]) | \(row.map { $0.map { String(format: "%.2f", $0) } ?? "N/A" }.joined(separator: " | ")) |")
    }
    return lines.joined(separator: "\n")
  }
}
